## Knapsack Problem:

Given weights and values of n items, put these items in a knapsack of capacity W to get the maximum total value in the knapsack. In other words, given two integer arrays val[0..n-1] and wt[0..n-1] which represent values and weights associated with n items respectively. Also given an integer W which represents knapsack capacity, find out the maximum value subset of val[] such that sum of the weights of this subset is smaller than or equal to W. You cannot break an item, either pick the complete item, or don’t pick it (0-1 property).

In [83]:
val = [60, 100, 120]
wt = [10, 20, 30]
W = 50
n = len(val)

In [84]:
def knapSack(W, wt, val, n):
    if n == 0 or W == 0 :                                           #Base case
        return 0                             
    if (wt[n-1] > W):                                        #If the weight of the Nth is more than the capacity of the knapsack
        return knapSack(W , wt , val , n-1)                         #Then the item cannot be included in the optimal solution
    else:                                                           #Else return the maximum of two cases:
        return max(val[n-1] + knapSack(W-wt[n-1] , wt , val , n-1), #(1)Nth item included
                   knapSack(W , wt , val , n-1))                    #(2)Nth item not included

In [85]:
knapSack(W, wt, val, n)

220

**Recursive Solution:**
<br>
> In this problem, we are given the weights and values of a "n" amount of items, as well as a knapsack of capacity "W".  We are asked to fill up the knapsack with the most possible value of items while remaining equal to or under the weight of the knapsack.  The solution to the knapsack problem is applied through a recursive function.  The first line of the knapsack function is the base case, where if N or W is equal to 0, we return a 0.  The next line says that if the weight of the Nth is more than the capacity of the knapsack, then the item cannot be included in the optimal solution.  If neither of these requirements are met, the function will return the maximum value value that can be obtained following two values: (1) Maximum value obtained by N-1 items and W (excluding Nth item), (2) Value of the Nth item plus maximum value obtained by N-1 items and W minus weight of the Nth item (including Nth item).

**Dynamic Solution:**
> One issue with the Recursive Function is that the function computes the same subfunctions over and over, making it very time consuming.  The run time for this function is O(2^n).  This can be addressed by adding a temporary array K, which allows us to solve the problem in a bottom up manner. From there, the function runs similarly to the recursive function.  The benefit of the dynamic function, however, is that it allows us to store each result into a table and then refer to the table for the optimal solution. The solution of this new dynamic function runs at O(nW) time, where n is the number of items and W is the weight.

In [88]:
def knapSack(W, wt, val, n): 
    K = [[0 for x in range(W+1)] for x in range(n+1)]                   #Initialize table K
    for i in range(n+1):                                                #Build K in a bottom-up manner
        for w in range(W+1): 
            if i==0 or w==0:                                           
                K[i][w] = 0
            elif wt[i-1] <= w:                                          
                K[i][w] = max(val[i-1] + K[i-1][w-wt[i-1]],  K[i-1][w]) 
            else: 
                K[i][w] = K[i-1][w] 
  
    return K[n][W] 

In [89]:
knapSack(W, wt, val, n)

220

In both algorithms, we arrive at an optimal solution of 220.